### PR TITLE
MANTA-4897 Overhaul rust zookeeper client MANTA-5215 Add multi-server support to rust zookeeper client MANTA-5217 Implement backoff stability threshold for rust zookeeper client

### PR DESCRIPTION
**Pull request by isaacdavis**
I'm putting this up now, even though there's more left to do. I'll file tickets for these things that are missing:
- [DONE] multiple-server connect string
- [DONE] stability threshold before resetting backoff
- test suite

the two "DONE" items are now a part of this pull request.

### CODE REVIEW

File: src/proto/decoder.rs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    replies: Arc<Mutex<HashMap<i32, ReplyRecord>>>,
+
+    // Global map of watches registered, indexed by path
+    watches: Arc<Mutex<HashMap<String, Vec<Watch>>>>,
+
+    //
+    // Global map of pending watches.
+    //
+    // Watches are only registered once we have confirmed that the operation
+    // that initiated the watch did not fail. Thust, we must stage watches here
```
Comment (kellymclaughlin) [inline src/proto/decoder.rs]: typo: Thust

File: src/proto/decoder.rs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    type Item = Result<i64, InternalError>;
+    //
+    // This is for unrecoverable errors.
+    // The Decoder trait requires this type to be convertible to an IoError, so
+    // we just use the IoError itself.
+    //
+    type Error = IoError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        fn wrap<T, E>(item: T) -> Result<Option<T>, E> {
```
Comment (kellymclaughlin) [inline src/proto/decoder.rs]: I'd suggest moving this to be just a top-level helper function since it doesn't operate in terms of anything inside the `Decoder` implementation.

File: src/proto/decoder.rs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        // Deserialize error
+        let zk_err: ZkError = buf.read_i32::<BigEndian>()?.into();
+        if zk_err != ZkError::Ok {
+            err = Some(zk_err);
+        }
+
+        //
+        // A watch has triggered.
+        //
+        if xid == WATCH_XID {
```
Comment (kellymclaughlin) [inline src/proto/decoder.rs]: The flow of this gets a little hard to follow due to the size of this outer `if/elseif/else` block. It might be nice to break out the handling of events from the control flow. Something like:
```
if xid == WATCH_XID {
    handle_watch_event(...)
} else if xid == HEARTBEAT_XID {
   handle_heartbeat(...)
} else {
    handle_request(...)
}

File: src/proto/request.rs [hunk 4]
```diff
@@ -120,73 +133,82 @@ pub(super) enum MultiHeader {
...
+        buf.put_u8(*self);
     }
 }
 
-impl WriteTo for str {
-    fn write_to<W: Write>(&self, mut writer: W) -> io::Result<()> {
-        writer.write_i32::<BigEndian>(self.len() as i32)?;
-        writer.write_all(self.as_ref())
+//
+// XXX is there a way to avoid implementing WriteTo for both String and &String?
```
Comment (kellymclaughlin) [inline src/proto/request.rs]: I don't think there is since the reference is treated as a separate type.

**Review by kellymclaughlin — COMMENTED** [review]
I haven't made it through all of this in detail yet, but I've got a few comments while I continue to go through it. This is really nice stuff!

File: src/client.rs [hunk 5]
```diff
@@ -0,0 +1,611 @@
...
+        &mut self,
+        rx: ReadHalf<TcpStream>,
+        enc_abort_handle: AbortHandle,
+    ) -> InternalError {
+        async fn recv_msg(
+            decoder: &mut FramedRead<ReadHalf<TcpStream>, ZkDecoder>,
+            sess_mgr: SessionManager,
+            log: Logger,
+        ) -> Result<(), InternalError> {
+            match decoder.next().await {
```
Comment (kellymclaughlin) [inline src/client.rs:399]: You could use an `if let` here to get rid of one level of nesting.

File: src/proto/decoder.rs (hunk 1 above)
Comment (isaacdavis) [reply src/proto/decoder.rs]: Fixed

File: src/proto/decoder.rs (hunk 2 above)
Comment (isaacdavis) [reply src/proto/decoder.rs]: With some of the other refactoring I decided that this function wasn't really necessary.

File: src/proto/decoder.rs (hunk 3 above)
Comment (isaacdavis) [reply src/proto/decoder.rs]: This is very true. I pulled the branches out into functions (except for the heartbeat since it's just a few lines).

File: src/proto/request.rs (hunk 4 above)
Comment (isaacdavis) [reply src/proto/request.rs]: Ah, well. I deleted the comment.

**Comment by isaacdavis** [comment]
Addressed first round of feedback.

I've also added multi-server support with ZkConnectString, borrowing the code from the primary resolver. This is MANTA-5215. I tested this functionality by connecting to a three-node zk cluster and killing various nodes, verifying that the client connects to a remaining node and that the backoff is incremented once for every time the client cycles through all the servers without connecting.

**Comment by isaacdavis** [comment]
Added backoff stability threshold. This adds the file `backoff.rs` and slightly simplifies some of the code in `session_manager.rs`. See MANTA-5217 for testing notes for this piece of functionality.

File: src/lib.rs [hunk 6]
```diff
@@ -93,283 +93,192 @@
...
-//! The futures in this crate expect to be running under a `tokio::Runtime`. In the common case,
-//! you cannot resolve them solely using `.wait()`, but should instead use `tokio::run` or
-//! explicitly create a `tokio::Runtime` and then use `Runtime::block_on`.
+//! The futures in this crate expect to be running under a `tokio::Runtime`.
+//! Within that context, you can `await` them.
 //!
-//! # A somewhat silly example
+//! # An example
 //!
 //! ```no_run
```
Comment (kellymclaughlin) [inline src/lib.rs:101]: I get an error here when running `cargo test`. I think it should be `rust,ignore` instead of `no_run` for now to avoid that error.

File: src/types/mod.rs [hunk 7]
```diff
@@ -1,3 +1,7 @@
+//
```
Comment (kellymclaughlin) [inline src/types/mod.rs:1]: I think nowadays it's more idiomatic to have a `type.rs` file at the same directory level and the `types` directory and then only include the specific submodules inside the `types` directory.

File: src/proto/mod.rs [hunk 8]
```diff
@@ -1,28 +1,8 @@
-use failure;
-use std::net::SocketAddr;
-use tokio;
-use tokio::prelude::*;
+//
```
Comment (kellymclaughlin) [inline src/proto/mod.rs:1]: Same as for my comment about the `types` module I would move this up one level to a `proto.rs` file.

File: src/types/mod.rs [hunk 9]
```diff
@@ -76,3 +109,73 @@ pub enum CreateMode {
     //  ^---- is it sequential?
     //   ^--- is it ephemeral?
 }
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum ZkConnectStringError {
```
Comment (kellymclaughlin) [inline src/types/mod.rs:114]: Add a comment for this type similar to the one above for `CreateMode` so that a description shows up in the generated docs.

**Review by kellymclaughlin — COMMENTED** [review]
I found a few missing links in the crate-level documentation. The following changes address it:
```
diff --git a/src/lib.rs b/src/lib.rs
index cb7be98..4cc1038 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -41,9 +41,9 @@
 //!
 //! ## Ephemeral nodes
 //!
-//! When you create a Znode, you also specify a [`CreateMode`]. Nodes that are created with
-//! [`CreateMode::Persistent`] are the nodes we have discussed thus far. They remain in the server
-//! until you delete them. Nodes that are created with [`CreateMode::Ephemeral`] on the other hand
+//! When you create a Znode, you also specify a [`CreateMode`](types/enum.CreateMode.html). Nodes that are created with
+//! [`CreateMode::Persistent`](types/enum.CreateMode.html#variant.Persistent) are the nodes we have discussed thus far. They remain in the server
+//! until you delete them. Nodes that are created with [`CreateMode::Ephemeral`](types/enum.CreateMode.html#variant.Ephemeral) on the other hand
 //! are special. These [ephemeral
 //! nodes](https://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#Ephemeral+Nodes) are
 //! automatically deleted by the server when the client that created them disconnects. This can be
@@ -53,8 +53,8 @@
 //!
 //! ## Watches
 //!
-//! In addition to the methods above, [`ZooKeeper::exists`], [`ZooKeeper::get_data`], and
-//! [`ZooKeeper::get_children`] also support setting
+//! In addition to the methods above, [`ZooKeeper::exists`](struct.ZooKeeper.html#method.exists), [`ZooKeeper::get_data`](struct.ZooKeeper.html#method.get_data), and
+//! [`ZooKeeper::get_children`](struct.ZooKeeper.html#method.get_children) also support setting
 //! "[watches](https://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#ch_zkWatches)" on
 //! a node. A watch is one-time trigger that causes a [`WatchedEvent`] to be sent to the client
 //! that set the watch when the state for which the watch was set changes. For example, for a
```

**Comment by isaacdavis** [comment]
Closing this PR; should it become relevant in the future please feel free to reopen and reassign as necessary.
